### PR TITLE
Allow particles to be applied to fields when particles aren't distributed throughout model.

### PR DESCRIPTION
**Pull request by Djneu**
These changes allow the cell/harmonic average, and nearest neighbor particle interpolators to apply the particles as a field even when there are areas with no particles. 

A few questions, as of now this assumes that any area with no particles should be given a value of zero. Would it be better to have this be user defined? Also, should I add a flag for this so that, generally, an assert would still be thrown unless this feature is specifically wanted?

### CODE REVIEW

**Comment by gassmoeller** [comment]
I like this feature, but it needs a bit of work. To your questions:
1. Do *you* need a value different from 0? If not, do not worry about things somebody else might need in the future. Often these things are never used and just lie around unused.
2. I would want to have this as a feature that is switched off by default. So by default the models should crash if there a no particles in a cell and its neighbors. I have two reasons for that: 1. To preserve the current behavior for old parameter files. 2. For most models that use the particles so far it is essential to have them in every cell (e.g. if they determine the viscosity in a cell). So continueing a run with empty cells will likely lead to bad results, and we should crash rather than continue. You can add a new input parameter (something like `Allow cells without particles`) and only ignore empty cells if this is set to `true`.

**Comment by Djneu** [comment]
Thanks for the comments! I've added a flag for allowing cells without particles that defaults to false.

File: include/aspect/particle/interpolator/cell_average.h [hunk 1]
```diff
@@ -52,6 +52,29 @@ namespace aspect
...
+          /**
+           * @copydoc Interface<dim>::parse_parameters()
+           **/
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          /**
+           * Generally, particles need to be distributed throughout the
```
Comment (gassmoeller) [inline include/aspect/particle/interpolator/cell_average.h]: This wording is not really clear to someone who is not familiar with how particles work. What about the following:
```
           * By default, every cell needs to contain particles to use this interpolator
           * plugin. If this parameter is set to true, cells are allowed to have no particles,
           * in which case the interpolator will return 0 for the cell's properties.
```

File: include/aspect/particle/interpolator/harmonic_average.h [hunk 2]
```diff
@@ -52,6 +52,29 @@ namespace aspect
...
+
+          /**
+           * @copydoc Interface<dim>::parse_parameters()
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          /**
```
Comment (gassmoeller) [inline include/aspect/particle/interpolator/harmonic_average.h]: Same comment.

File: include/aspect/particle/interpolator/nearest_neighbor.h [hunk 3]
```diff
@@ -53,6 +53,29 @@ namespace aspect
...
+
+          /**
+           * @copydoc Interface<dim>::parse_parameters()
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          /**
```
Comment (gassmoeller) [inline include/aspect/particle/interpolator/nearest_neighbor.h]: and here

File: source/particle/interpolator/cell_average.cc [hunk 4]
```diff
@@ -115,17 +115,65 @@ namespace aspect
...
+                // Assume property is zero for any areas with no particles
+                else if (allow_cells_without_particles && non_empty_neighbors == 0)
+                  cell_properties[i] = 0;
+              }
+
           }
 
         return std::vector<std::vector<double> > (positions.size(),cell_properties);
       }
+
```
Comment (gassmoeller) [inline source/particle/interpolator/cell_average.cc]: We usually use 3 empty lines between functions, please change.

File: source/particle/interpolator/cell_average.cc [hunk 5]
```diff
@@ -115,17 +115,65 @@ namespace aspect
...
+      void
+      CellAverage<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+          prm.enter_subsection("Particles");
+          {
+            prm.declare_entry ("Allow cells without particles", "false",
+                               Patterns::Bool (),
+                               "Generally, particles need to be distributed throughout the "
```
Comment (gassmoeller) [inline source/particle/interpolator/cell_average.cc]: Same text change here.

File: source/particle/interpolator/cell_average.cc [hunk 6]
```diff
@@ -115,17 +115,65 @@ namespace aspect
...
+      template <int dim>
+      void
+      CellAverage<dim>::parse_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+          prm.enter_subsection("Particles");
+          {
+            allow_cells_without_particles = prm.get_bool("Allow cells without particles");
+
```
Comment (gassmoeller) [inline source/particle/interpolator/cell_average.cc]: Remove empty line. Also in the other plugins.

File: source/particle/interpolator/harmonic_average.cc [hunk 7]
```diff
@@ -118,17 +118,64 @@ namespace aspect
...
+                // Assume property is zero for any areas with no particles
+                else if (allow_cells_without_particles && non_empty_neighbors == 0)
+                  cell_properties[i] = 0;
+              }
+
           }
 
         return std::vector<std::vector<double> > (positions.size(),cell_properties);
       }
+
```
Comment (gassmoeller) [inline source/particle/interpolator/harmonic_average.cc]: three empty lines here and at every other gap between functions.

**Comment by Djneu** [comment]
Okay, I believe I addressed all the styling issues.

**Comment by gassmoeller** [comment]
Ready to merge once the tester is finished.

**Pull request merged**